### PR TITLE
perf: cost-model based persistent kernel load-balancing scheduler

### PR DESCRIPTION
**Pull request by yzh119**
This PR implements the cost-model based persistent kernel load-balancing scheduler.
In our previous design, we use binary search to search for the best chunk size. However, this method is sub-optimal because:
1. chunk size doesn't have to be a fixed value.
2. there are still some possible wave quantization, even after scheduling.
3. binary search only splits long sequence, but didn't pack short sequences.

This PR fixes these issues, and there will be zero wave quantization after this PR.

### CODE REVIEW

File: python/flashinfer/utils.py [hunk 1]
```diff
@@ -211,17 +211,19 @@ def register_custom_op(
...
-        return lambda x: x
-    return torch.library.custom_op(
-        name, fn, mutates_args=mutates_args, device_types=device_types, schema=schema
-    )
+    # if TorchVersion(torch_version) < TorchVersion("2.4"):
+    #     return lambda x: x
+    # return torch.library.custom_op(
+    #     name, fn, mutates_args=mutates_args, device_types=device_types, schema=schema
+    # )
+    return lambda x: x
```
Comment (yzh119) [inline python/flashinfer/utils.py:219]: Will add `custom_op` back after we figured out the performance degradation @abcdabcd987 .

**Comment by zhyncs** [comment]
Shoutout to yzh for god-like work.

**Comment by merrymercy** [comment]
Can we merge this soon?

**Comment by yzh119** [comment]
@merrymercy  yes but this branch is forsaken, I'll create another PR.

**Comment by yzh119** [comment]
Moved to another PR for development.

**Pull request closed without merging**